### PR TITLE
feat(ast_codegen): add `ImplGetSpanGenerator`.

### PR DESCRIPTION
**Pull request by rzvxa**
This one is ready to replace the handwritten module.

### CODE REVIEW

**Comment by rzvxa** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3852?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3852** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3852?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#3815** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3815?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3840** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3840?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_ast/src/lib.rs [hunk 1]
```diff
@@ -17,11 +17,15 @@ mod ast_builder;
 mod ast_impl;
 mod ast_kind;
 pub mod precedence;
-mod span;
+// mod span;
```
Comment (overlookmotel) [inline crates/oxc_ast/src/lib.rs:20]: This line can be removed.

**Comment by overlookmotel** [comment]
A couple of small things:

Can we add `#[inline]` on all the `fn span`s for structs? Probably they'll be inlined anyway, but we should make sure. NB: Probably shouldn't `#[inline]` the functions for enums, as they're more complex.

Is there any way to add line breaks between the `impl`s so the generated code is easier to read? I remember seeing other codegens use a crate which pretty-prints `syn` output (though I can't remember what it's called).

**Comment by rzvxa** [comment]
@overlookmotel You can go ahead and review it. Let me know if there is anything ambiguous that needs to be addressed.

**Comment by overlookmotel** [comment]
> @overlookmotel You can go ahead and review it. Let me know if there is anything ambiguous that needs to be addressed.

To be honest, I don't follow everything you're doing here. As discussed, I think it'd be simpler if we purely used the `DefType` types as the source for generation. But if the tests pass, let's go with it!

**Comment by overlookmotel** [comment]
@Boshen unless you have any comments, this one is ready to merge I think.

**Comment by overlookmotel** [comment]
As discussed in https://github.com/oxc-project/oxc/issues/3900#issuecomment-2188460531, it'd be preferable in my view to remove the "special case" code for `BindingPattern` from the codegen, and use a `#[span]` attr in AST type to indicate what needs to happen to codegen.

But am really keen to start getting this stack merged, so let's not change it now. Merge without further churn, and then go back and change it in a later PR.

**Review by overlookmotel — APPROVED** [review]

**Pull request merged**